cub/block/block_run_length_decode.cuh

```diff
@@ -0,0 +1,435 @@
+  {
+    // Compute the offset for the beginning of each run
+    RelativeOffsetT run_offsets[RUNS_PER_THREAD];
+#pragma unroll
+    for (uint32_t i = 0; i < RUNS_PER_THREAD; i++)
+    {
+      run_offsets[i] = run_lengths[i];
+    }
+    RelativeOffsetT decoded_size_aggregate;
+    RunOffsetScanT(temp_storage.offset_scan).ExclusiveSum(run_offsets, run_offsets, decoded_size_aggregate);
+    total_decoded_size = decoded_size_aggregate;
+
+    // Ensure the prefix scan's temporary storage can be reused (may be superfluous, but depends on scan implementaiton)
+    __syncthreads();
+
+    // Keep the runs' unique items and the offsets of each run's beginning in the temporary storage
+    RunOffsetT thread_dst_offset = linear_tid * RUNS_PER_THREAD;
+#pragma unroll
+    for (uint32_t i = 0; i < RUNS_PER_THREAD; i++)
+    {
+      temp_storage.runs.unique_items[thread_dst_offset] = unique_items[i];
+      temp_storage.runs.run_offsets[thread_dst_offset]  = run_offsets[i];
+      thread_dst_offset++;
+    }
+
+    // Write the runs' total decoded size into the last element
+    if (linear_tid == 0)
+    {
+      temp_storage.runs.run_offsets[BLOCK_RUNS] = total_decoded_size;
+    }
+
+    // Ensure run offsets have been writen to shared memory
+    __syncthreads();
+  }
+
+  /**
+   * @brief Run-length decodes the runs previously passed via a call to Init(...) and returns the run-length decoded
+   * items in a blocked arrangement to \p decoded_items. If the number of run-length decoded items exceeds the
+   * run-length decode buffer (i.e., <b>DECODED_ITEMS_PER_THREAD * BLOCK_THREADS</b>), only the items that fit within
+   * the buffer are returned. Subsequent calls to <b>RunLengthDecode</b> adjusting \p from_decoded_offset can be
+   * used to retrieve the remaining run-length decoded items.
+   *
+   * @param from_decoded_offset If invoked with from_decoded_offset that is larger than total_decoded_size results in
+   * undefined behavior.
+   */
+  template <typename UserRelativeOffsetT>
+  __device__ __forceinline__ void RunLengthDecode(UniqueItemT (&decoded_items)[DECODED_ITEMS_PER_THREAD],
+                                                  UserRelativeOffsetT (&item_offsets)[DECODED_ITEMS_PER_THREAD],
+                                                  int32_t from_decoded_offset = 0)
+  {
+    // Populate run-length decode buffer with continuation-of-current-run items
+    InitDecodeBuffer(temp_storage.runs.decode_buffer);
+
+    // Ensure that the continuation-of-current-run items have been written to shared memory
+    __syncthreads();
+
+    // Scatter the beginning of each run into the run-length decode buffer
+    RunOffsetT run_offset = linear_tid;
+#pragma unroll
+    for (uint32_t i = 0; i < RUNS_PER_THREAD; i++)
+    {
+      // Fetch this thread's run and subsequently check whether it falls into this pass' run-length decode buffer
+      int32_t run_target_idx      = temp_storage.runs.run_offsets[run_offset];
```

Comment:
The type of `temp_storage.runs.run_offsets` is `RelativeOffsetT`, can we change the type of `next_run_target_idx` to `RelativeOffsetT`?

Reply:
Has become obsolete with the switch to binary search-based method.